### ISSUE TITLE
Timestamp not preserved when rsyncing to SMB share on macOS

### ISSUE THREAD

**Issue by iceboundflame** [WayneD/rsync#479]
I am using rsync 3.2.7, installed via Homebrew on macOS Ventura 13.2.1. `rsync --archive` does not set the timestamp correctly when the destination is a mounted SMB share.

When syncing to a local filesystem directory, the timestamp is copied correctly.
Also when I switch to the builtin macOS version 2.6.9, it does work correctly as well.

Create test file:
```
$ echo helloworld > test/asdf
$ touch -t 202301010000 test/asdf
$ ls -l test
total 8
-rw-r--r--  1 dliu  staff  11 Jan  1 00:00 asdf
```

Sync to mount, timestamp not preserved:
```
$ /opt/homebrew/Cellar/rsync/3.2.7_1/bin/rsync -avr --itemize-changes test /Volumes/home/
sending incremental file list
.d...p...... test/
>f..tp...... test/asdf

sent 145 bytes  received 39 bytes  368.00 bytes/sec
total size is 11  speedup is 0.06

$ ls -l /Volumes/home/test
total 8
-rwx------  1 dliu  staff  11 May  7 16:47 asdf
```

Broken version details:
```
$ /opt/homebrew/Cellar/rsync/3.2.7_1/bin/rsync --version
rsync  version 3.2.7  protocol version 31
Copyright (C) 1996-2022 by Andrew Tridgell, Wayne Davison, and others.
Web site: https://rsync.samba.org/
Capabilities:
    64-bit files, 64-bit inums, 64-bit timestamps, 64-bit long ints,
    socketpairs, symlinks, symtimes, hardlinks, hardlink-specials,
    hardlink-symlinks, IPv6, atimes, batchfiles, inplace, append, ACLs,
    xattrs, optional secluded-args, iconv, no prealloc, stop-at, crtimes,
    file-flags
Optimizations:
    no SIMD-roll, no asm-roll, openssl-crypto, no asm-MD5
Checksum list:
    xxh128 xxh3 xxh64 (xxhash) md5 md4 sha1 none
Compress list:
    zstd lz4 zlibx zlib none
Daemon auth list:
    sha512 sha256 sha1 md5 md4

rsync comes with ABSOLUTELY NO WARRANTY.  This is free software, and you
are welcome to redistribute it under certain conditions.  See the GNU
General Public Licence for details.
```

**Comment by botsmack** [comment]
I think I'm running into this issue as well. I noticed it while syncing data from my iMac using rsync 3.2.7 (installed via Homebrew) on macOS Ventura 13.2.1. I am mounting an smb share and running the following command:

```
rsync -rltvzhP ~/folder/data/ /Volumes/tank/differentfolder/olddata
```

Running the above command successively will re-sync everything I've just transferred. The modified date is not being being honored or carried over, but rather being updated to the datetime that the current rsync command is carried out. The details of my SMB connection as reported from my Mac to an Unraid share:

```
SMB_NEGOTIATE                 SMBV_NEG_SMB1_ENABLED
SMB_NEGOTIATE                 SMBV_NEG_SMB2_ENABLED
SMB_NEGOTIATE                 SMBV_NEG_SMB3_ENABLED
SMB_VERSION                   SMB_3.1.1
SMB_ENCRYPT_ALGORITHMS        AES_128_CCM_ENABLED
SMB_ENCRYPT_ALGORITHMS        AES_128_GCM_ENABLED
SMB_ENCRYPT_ALGORITHMS        AES_256_CCM_ENABLED
SMB_ENCRYPT_ALGORITHMS        AES_256_GCM_ENABLED
SMB_CURR_ENCRYPT_ALGORITHM    AES-128-GCM
SMB_SHARE_TYPE                DISK
SIGNING_SUPPORTED             TRUE
EXTENDED_SECURITY_SUPPORTED   TRUE
UNIX_SUPPORT                  TRUE
LARGE_FILE_SUPPORTED          TRUE
OS_X_SERVER                   TRUE
DFS_SUPPORTED                 TRUE
FILE_LEASING_SUPPORTED        TRUE
MULTI_CREDIT_SUPPORTED        TRUE
ENCRYPTION_REQUIRED           TRUE
```

**Comment by rdcyuxi** [comment]
I have the same issue with rsync 3.3.0 in /opt/homebrew/bin installed by homebrew, alternatively OS shipped rsync 2.6.9 in /usr/bin works well.

Another alternative is to share server volume with AFP/Netatalk. Both rsync work well.

**Comment by Pkaster** [comment]
I'm having the exact same issue with:
rsync  version 3.4.1  protocol version 32
Server: Synology DS with DSM 7.2.2

I don't know if it is an issue with rsync, the smb client or the smb server on the Synology (or all the three at the same time), but I could solve it by enabling `Synchronize data to drive immediately upon SMB client request` in DSM as explained here: https://serverfault.com/questions/1123027/rsync-is-not-preserving-file-timestamps-with-the-a-option-when-writing-to-cifs

**Comment by iceboundflame** [comment]
Yeah, still seeing this with 3.4.1 and a TrueNAS file server. Here is a workaround Claude found that worked for me:

> Workaround: add --fsync to your rsync command:
> 
> rsync -a --fsync test /Volumes/home/
> 
> This fixes the problem reliably on macOS → SMB. Alternatively, if you control the server, enabling strict sync there also fixes it (on Synology DSM: "Synchronize data to drive immediately upon SMB client request", per @Pkaster above; Samba: strict sync = yes, the default since 4.7.0).
> 
> Why this happens (for posterity):
> 
> This turns out not to be an rsync bug, but a flush-ordering problem between rsync, the macOS SMB client, and the server's write cache.
> 
> For each file, rsync does: write temp file → close → set mtime (setattrlist()/utimes()) → rename into place. On a local filesystem that's fine. On an SMB mount it's unsafe, because both Apple's SMB client (aggressive write-behind caching, deferred close) and the server (write cache) may not commit the file's data until after rsync has already set the timestamp. At the protocol level the order becomes:
> 
> 1. WRITE — data buffered, not yet destaged to disk
> 2. SET_INFO — rsync sets the source mtime (this works, briefly)
> 3. FLUSH/CLOSE — the server finally writes the data to disk, and updates mtime to "now", stomping the value rsync just set
> 
> --fsync makes rsync fsync each file before closing it and setting the timestamp. That forces an SMB FLUSH, so the server destages the data (and does its implicit mtime bump) first; rsync's timestamp then lands last and sticks. The server-side strict-sync setting fixes the same race from the other end.
> 
> This is a known SMB footgun: the Linux kernel CIFS client guards against it internally (cifs: flush before set-info if we have writeable handles, 2019) — which is why Linux clients don't show this. Apple's smbfs has no equivalent guard, so every rsync-like tool on macOS is exposed: this reproduces with Homebrew rsync 3.2.7 through 3.4.1 and also with Apple's openrsync shipped in macOS 15.4+.
> 
> Two consequences worth noting:
> 
> - --modify-window does not help — the destination mtime becomes the transfer time, not slightly-off, which is also why every subsequent run re-transfers everything.
> - --fsync costs some throughput (one flush round-trip per file, most noticeable with many small files), but for repeated syncs it pays for itself by eliminating the spurious full re-transfers.

**Comment by steadytao** [comment]
I do not think we have enough evidence yet to conclude that this is definitively outside rsync. A packet trace or equivalent comparison between the failing run, `--fsync` and Apples older rsync would help confirm the exact SMB operation ordering before we close this.

**Comment by iceboundflame** [comment]
FWIW I tried the old rsync 2.6.9 again and it was not syncing times correctly anymore either. Too many things have changed in my setup in the last 3 years to say why.

See also: https://forums.unraid.net/topic/192886-rsync-from-macos-over-smb-doesnt-preserve-modification-time/ - this may be where the --fsync suggestion comes from.
